packages/server/src/models/Locker.ts

```diff
@@ -0,0 +1,106 @@
+import {ERRORS} from '../constants'
+
+export type RequestRelease = () => Promise<void> | void
+
+// The Locker interface defines methods for implementing a locking mechanism.
+// This is crucial for ensuring exclusive access to uploads and their metadata.
+// Following TUS recommendations, it's important to cancel locks from previous requests
+// to avoid holding locks for too long and to manage half-open TCP connections effectively.
+// The lock method includes a cancel callback to facilitate the cancellation of a request that previously acquired the lock.
+export interface Locker {
+  lock(id: string, cancelReq: RequestRelease): Promise<void>
+  unlock(id: string): Promise<void>
+}
+
+export interface MemoryLockerOptions {
+  acquireLockTimeout: number
+}
+
+class Lock {
+  public requestRelease?: RequestRelease
+}
+
+// MemoryLocker is an implementation of the Locker interface, maintaining locks in memory.
+// It ensures exclusive access to upload resources by managing locks for each upload.
+// The lock() method ensures that any previously held lock is released before acquiring a new one.
+// Lock acquisition attempts will timeout based on the specified 'acquireLockTimeout' duration.
+export class MemoryLocker implements Locker {
+  private locks = new Map<string, Lock>()
+  protected timeout: number
+
+  constructor(options?: MemoryLockerOptions) {
+    this.timeout = options?.acquireLockTimeout ?? 1000 * 30
+  }
+
+  async lock(id: string, requestRelease: RequestRelease): Promise<void> {
+    const abortController = new AbortController()
+    const lock = await Promise.race([
+      this.waitTimeout(abortController.signal),
+      this.acquireLock(id, abortController.signal),
+    ])
+
+    abortController.abort()
+
+    if (!lock) {
+      throw ERRORS.ERR_LOCK_TIMEOUT
+    }
+    lock.requestRelease = requestRelease
+  }
+
+  protected async acquireLock(id: string, signal: AbortSignal): Promise<Lock | void> {
+    if (signal.aborted) {
+      return
+    }
+
+    const lock = this.locks.get(id)
+
+    if (!lock) {
+      const lock = new Lock()
+      this.locks.set(id, lock)
+      return lock
+    }
+
+    await lock.requestRelease?.()
+
+    const potentialNewLock = this.locks.get(id)
```

Comment:
Is this code around `potentialNewLock` actually necessary? Isn't the acquiring of the lock handled by the `this.acquireLock` call a few lines below?

Reply:
Yes good point!